python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py

```diff
@@ -243,3 +254,84 @@ def get_unique_asset_id(
         if repository_identifier
         else f"{asset_key.to_string()}"
     )
+
+
+def get_freshness_info(
+    instance: DagsterInstance,
+    asset_key: AssetKey,
+    asset_graph: AssetGraph,
+    freshness_policy: FreshnessPolicy,
+    batch_loader: BatchMaterializationLoader,
+):
+    from ..schema.asset_graph import GrapheneAssetFreshnessInfo
+
+    current_time = datetime.datetime.now(tz=datetime.timezone.utc)
+
+    def _convert_timestamp(timestamp: Optional[float]) -> Optional[datetime.datetime]:
+        if timestamp is None:
+            return None
+        return datetime.datetime.fromtimestamp(timestamp, tz=datetime.timezone.utc)
+
+    # must use get_event_records instead of using the batch loader because we need the
+    # storage_id on the record
+    latest_record = get_latest_materialization_record(instance=instance, asset_key=asset_key)
+
+    # for now, we always set the relevant upstream keys to the root assets
+    relevant_upstream_keys = asset_graph.get_roots(asset_key)
+
+    used_upstream_materialization_times = get_used_upstream_materialization_times_for_record(
+        instance=instance,
+        asset_graph=asset_graph,
+        upstream_keys=relevant_upstream_keys,
+        record=latest_record,
+    )
+
+    # not all FreshnessPolicies require this information, so don't fetch it if not necessary
+    if isinstance(freshness_policy, MaximumLatencyFreshnessPolicy):
+
+        # at the current moment, the most recent data available for each of the relevant upstream
+        # asset keys
+        current_times = {
+            upstream_key: _convert_timestamp(
+                batch_loader.get_latest_materialization_for_asset_key(upstream_key).timestamp
+            )
+            for upstream_key in relevant_upstream_keys
+        }
+
+        # same as above, but for the latest data that was available when we actually materialized
+        # the latest record
+        original_times = {
+            upstream_key: _convert_timestamp(
+                get_latest_materialization_record(
```

Comment:
Would we be able to use the `TickInstanceQueryer` (maybe rename it) so we can cache these?